feature/experimental/gribi/ate_tests/supervisor_failure/Supervisor_Failure_test.go

```diff
@@ -0,0 +1,375 @@
+	i2 := top.AddInterface(atePort2.Name).WithPort(p2)
+	i2.IPv4().
+		WithAddress(atePort2.IPv4CIDR()).
+		WithDefaultGateway(dutPort2.IPv4)
+
+	return top
+}
+
+// createTrafficFlow generates traffic flow from source network to
+// destination network via srcEndPoint to dstEndPoint and checks for
+// packet loss.
+func createTrafficFlow(t *testing.T, ate *ondatra.ATEDevice, top *ondatra.ATETopology, srcEndPoint, dstEndPoint *ondatra.Interface) *ondatra.Flow {
+	ethHeader := ondatra.NewEthernetHeader()
+	ipv4Header := ondatra.NewIPv4Header()
+	ipv4Header.DstAddressRange().
+		WithMin("203.0.113.0").
+		WithMax("203.0.113.254").
+		WithCount(250)
+
+	flow := ate.Traffic().NewFlow("Flow").
+		WithSrcEndpoints(srcEndPoint).
+		WithDstEndpoints(dstEndPoint).
+		WithHeaders(ethHeader, ipv4Header)
+
+	return flow
+
+}
+
+// Function to send traffic
+func sendTraffic(t *testing.T, ate *ondatra.ATEDevice, flow *ondatra.Flow) {
+	t.Logf("Starting traffic")
+	ate.Traffic().Start(t, flow)
+}
+
+// Function to verify traffic
+func verifyTraffic(t *testing.T, ate *ondatra.ATEDevice, flow *ondatra.Flow) {
+	flowPath := ate.Telemetry().Flow(flow.Name())
+	if got := flowPath.LossPct().Get(t); got > 0 {
+		t.Errorf("LossPct for flow %s got %g, want 0", flow.Name(), got)
+	} else {
+		t.Logf("Traffic flows fine from ATE-port1 to ATE-port2")
+	}
+}
+
+// Function to stop traffic
+func stopTraffic(t *testing.T, ate *ondatra.ATEDevice) {
+	t.Logf("Stopping traffic")
+	ate.Traffic().Stop(t)
+}
+
+// testArgs holds the objects needed by a test case.
+type testArgs struct {
+	ctx     context.Context
+	clientA *gribi.Client
+	dut     *ondatra.DUTDevice
+	ate     *ondatra.ATEDevice
+	top     *ondatra.ATETopology
+}
+
+// routeInstall configures a IPv4 entry through clientA. Ensure that the entry via ClientA
+// is active through AFT Telemetry.
+func routeInstall(ctx context.Context, t *testing.T, args *testArgs) {
+	// Add an IPv4Entry for 203.0.113.0/24 pointing to ATE port-2 via gRIBI-A,
+	// ensure that the entry is active through AFT telemetry
+	t.Logf("Add an IPv4Entry for %s pointing to ATE port-2 via gRIBI-A", ateDstNetCIDR)
+	args.clientA.AddNH(t, nhIndex, atePort2.IPv4, instance, fluent.InstalledInRIB)
+	args.clientA.AddNHG(t, nhgIndex, map[uint64]uint64{nhIndex: 1}, instance, fluent.InstalledInRIB)
+	args.clientA.AddIPv4(t, ateDstNetCIDR, nhgIndex, instance, "", fluent.InstalledInRIB)
+
+	// Verify the entry for 203.0.113.0/24 is active through AFT Telemetry.
+	ipv4Path := args.dut.Telemetry().NetworkInstance(instance).Afts().Ipv4Entry(ateDstNetCIDR)
+	if got, want := ipv4Path.Prefix().Get(t), ateDstNetCIDR; got != want {
+		t.Errorf("ipv4-entry/state/prefix got %s, want %s", got, want)
+	} else {
+		t.Logf("ipv4-entry entry found for %s before controller switchover..", got)
+	}
+}
+
+// findSecondaryController finds out primary and secondary controllers
+func findSecondaryController(t *testing.T, dut *ondatra.DUTDevice, controllers []string) (string, string) {
```

Comment:
We will move this to a lib so that we do not have same copy for switchover test.